Added support for little endian data for patches

Tupelov opened the pull request:
Allows users to use the leshort, leword, and ledouble datatypes inside their patches. Using these datatypes ensures that values in the memory view appear the same as the values in the patch.
This was made with easing patch development in mind as now developers can make patches without having to reverse endianness. This is especially useful when using programs like ghidra who display an instructions bytes in little endian.
![image](https://user-images.githubusercontent.com/64274155/109246155-baafd600-77af-11eb-89c0-3f78829ce07a.png)
![image](https://user-images.githubusercontent.com/64274155/109246288-f0ed5580-77af-11eb-8263-e1ff5c7496b5.png)

AmFobes commented on pcsx2/Patch.h:
```diff
@@ -128,3 +131,8 @@ extern void PatchesVerboseReset();
+//Swaps endianess of InputNum
+//ex. 01020304 -> 04030201
+//Could be made more portable if we remove the patch_data_type bar
+extern u64 SwapEndian(u64 InputNum, patch_data_type);
```
Add a space after `//` for consistency with other comments

Do you also perhaps mean `var` instead of `bar` in your comment?

Tupelov replied:
Fixed in https://github.com/PCSX2/pcsx2/pull/4255/commits/006e946060582292f3e9b935f97268407d796010

AmFobes reviewed (commented):
Functionally seems to work fine.

Nice addition, can't tell you how many times I've forgotten that pnach requires a different endianness when Ghidra is used.

refractionpcsx2 commented on pcsx2/Patch_Memory.cpp:
```diff
@@ -420,6 +420,23 @@ void _ApplyPatch(IniPatch *p)
+			break;
+
+		case DOUBLE_LE_T:
+			u64 memle;
```
This is gonna bitch about scope probably, so you'll want to put this case in brackets.

Tupelov replied:
Looking back it now, I realize that the if statement is probably to check if the memory has already been written so Ill go fix that.
And now I know why I had scope issues with the switch statements, I thought each case was its own scope.

Tupelov replied:
Ok I addressed the scope and comparison issue in https://github.com/PCSX2/pcsx2/pull/4255/commits/501f6199cf50a5b004d00d2ed4b87e201965c813

refractionpcsx2 commented on pcsx2/Patch_Memory.cpp:
```diff
@@ -449,3 +466,18 @@ void _ApplyPatch(IniPatch *p)
+	}
+	InputNum = (InputNum & 0x00FF00FF00FF00FF) << 8 | (InputNum & 0xFF00FF00FF00FF00) >> 8;   //Swaps 1 byte
+	return InputNum;
+}
```
is the > 1 byte swapping necessary? I've never seen that be a thing.

refractionpcsx2 commented on pcsx2/Patch_Memory.cpp:
```diff
@@ -449,3 +466,18 @@ void _ApplyPatch(IniPatch *p)
+	}
+	InputNum = (InputNum & 0x00FF00FF00FF00FF) << 8 | (InputNum & 0xFF00FF00FF00FF00) >> 8;   //Swaps 1 byte
+	return InputNum;
+}
```
Please put spaces between // and the comments (like you did with // Swaps 2 bytes)

Tupelov replied:
Fixed in https://github.com/PCSX2/pcsx2/pull/4255/commits/006e946060582292f3e9b935f97268407d796010

Tupelov commented:
Alright, I fixed the issues, in short I fixed the comments, made the SwapEndian function more portable, fixed some scope issues, and finally I fixed a memory check issue.

refractionpcsx2 commented on pcsx2/Patch_Memory.cpp:
```diff
@@ -394,6 +394,8 @@ void _ApplyPatch(IniPatch *p)
 		switch (p->type)
 		{
+		u64 mem;
+		u64 ledata;
```
Can you move these two to the beginning of the _ApplyPatch() function and initialise them to 0 please.
I don't know how declaring them in a switch like this is valid, it shouldn't be lol.

Tupelov replied:
Alright, will do, sorry about all these switch issues lol I don't usually use them. I think the compiler didn't complain because I put the variables within the brackets of the switch case

Tupelov replied:
Addressed in https://github.com/PCSX2/pcsx2/pull/4255/commits/94b198085c3679ae3cf55c45cc4bb9bf20493fd0

refractionpcsx2 commented on pcsx2/Patch_Memory.cpp:
```diff
@@ -420,6 +423,25 @@ void _ApplyPatch(IniPatch *p)
+		case WORD_LE_T:
+			ledata = SwapEndian(p->data, 32);
+			if (memRead32(p->addr) != (u32)ledata)
+				memWrite32(p->addr, (u32) ledata );
```
remove the spacing on the ledata parameter please :)  so it should be memWrite32(p->addr, (u32)ledata);

Also do the same for the memWrite64 below.  (last thing, I promise lol)

Tupelov commented:
Alright, hopefully this should be enough

refractionpcsx2 commented on pcsx2/Patch_Memory.cpp:
```diff
@@ -420,6 +423,25 @@ void _ApplyPatch(IniPatch *p)
+		case WORD_LE_T:
+			ledata = SwapEndian(p->data, 32);
+			if (memRead32(p->addr) != (u32)ledata)
+				memWrite32(p->addr,(u32)ledata);
```
space between , and  (u32)
memWrite32(p->addr, (u32)ledata);

refractionpcsx2 commented on pcsx2/Patch_Memory.cpp:
```diff
@@ -420,6 +423,25 @@ void _ApplyPatch(IniPatch *p)
+			ledata = SwapEndian(p->data, 64);
+			memRead64(p->addr, &mem);
+			if (mem != ledata)
+				memWrite64(p->addr,ledata);
```
space between , and ledata
memWrite64(p->addr, ledata);

refractionpcsx2 commented on pcsx2/Patch_Memory.cpp:
```diff
@@ -387,13 +387,17 @@ void handle_extended_t(IniPatch *p)
 	case CPU_EE:
 		switch (p->type)
 		{
+		
```
Where'd this line come from?

Tupelov replied:
pain

Tupelov commented:
Kind of off-topic but since were here, is there any ETA for pcsx2 1.7? or at least a 1.6.1? Hopefully soon because it would be kinda dumb to write patches that require you to build from source to use lol

refractionpcsx2 commented:
> Kind of off-topic but since were here, is there any ETA for pcsx2 1.7? or at least a 1.6.1? Hopefully soon because it would be kinda dumb to write patches that require you to build from source to use lol

we don't release 1.7 (well, we do daily), but the next stable release will be 1.8 which we're planning on doing later this year

Tupelov commented:
Thanks for the reply, I was curious. Although I've heard pcsx2 is going under a major rewrite when it comes to configs. I have some feature ideas but I don't want to jump the gun and then have to rewrite for it to work with the new implementation.

refractionpcsx2 commented:
Yes probably hold off on that for a moment, we're rewriting the config stuff in YAML and merging all the plugins etc, so quite heavy bits of work going on at the moment.

Also can you fix those other 2 things please :)

Tupelov commented:
> 
> 
> Yes probably hold off on that for a moment, we're rewriting the config stuff in YAML and merging all the plugins etc, so quite heavy bits of work going on at the moment.
> 
> Also can you fix those other 2 things please :)

Alright thanks, I somehow missed those over 2 things lol, I addressed them in https://github.com/PCSX2/pcsx2/pull/4255/commits/6d42c7354c3e2f7d668e5152ffc84c75647a0582

refractionpcsx2 commented:
Ok PR looks good now, will merge it once the builds complete :)

Tupelov commented:
Alright, thanks for guidance :)
Have a nice day.

refractionpcsx2 commented:
> Alright, thanks for guidance :)
> Have a nice day.

Thank you, you too
